edgedb-client/src/builder.rs

```diff
@@ -322,6 +329,171 @@ impl fmt::Display for DisplayAddr<'_> {
     }
 }
 
+impl DSN {
+    fn new(dsn: &str) -> Result<Self, Error> {
+        let admin = dsn.starts_with("edgedbadmin://");
+        if !dsn.starts_with("edgedb://") && !admin {
+            return Err(ClientError::with_message(format!(
+                "String {:?} is not a valid DSN",
+                dsn,
+            )));
+        };
+        let url = url::Url::parse(dsn).map_err(|e| {
+            ClientError::with_source(e).context(format!("cannot parse DSN {:?}", dsn))
+        })?;
+        let query = url
+            .query_pairs()
+            .into_owned()
```

Comment:
I think you could do this without making them owned. Just chosen a bit wrong set of abstractions here. Kinda this:
```
let url = Url::parse(..)?;
let query = url.query_pairs().collect::<HashMap<_, _>();
let host = get_value(url, query, "host", ...);
let port = get_value(url, query, "port", ...);
```
I.e. unpack values while URL is still on stack.

This can be further wrapped into a structure with a lifetime:
```
struct DsnHelper<'a> {
    url: &'a Url,
    query: &'a HashMap<..>,
}
let url = Url::parse(..)?;
let dsn = DsnHelper::from_url(url)?;
let host = dsn.get_host()?;
let port = dsn.get_port()?;
```

But it's unlikely that this will be a performance issue, so it's very minor thing here.